### PR TITLE
pgpool: 3.4.14 -> 4.0.5

### PR DESCRIPTION
**Pull request by takeda**
###### Motivation for this change

I needed to use latest version of pgpool (4.0.5) and wanted to contribute my change back.
This change also patches the code to allow providing an absolute path for pool_passwd. Originally pgpool is looking for the pool_passwd in the same directory where configuration is located. Having it accept an absolute path gives more flexibility.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS (pgbouncer doesn't compile on OS X without an extra effort so I didn't bother)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) (no tests that I'm aware of)
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) (was getting an error about stdenv: `error: cannot auto-call a function that has an argument without a default value ('stdenv')`)
- [ ] Ensured that relevant documentation is up to date (this is just a package and I don't believe there's a documentation beyond the meta data)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @ (no idea who is a maintainer for it)

### CODE REVIEW

File: pkgs/servers/sql/pgpool/pgpool.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ 	pool_semaphore_create(MAX_NUM_SEMAPHORES);
+diff --git a/src/watchdog/wd_lifecheck.c b/src/watchdog/wd_lifecheck.c
+index a24cad85..ae29554e 100644
+--- a/src/watchdog/wd_lifecheck.c
++++ b/src/watchdog/wd_lifecheck.c
+@@ -26,6 +26,7 @@
+ #include <ctype.h>
+ #include <time.h>
+ #include <string.h>
++#include <stdint.h>
```
Comment (danbst) [inline pkgs/servers/sql/pgpool/pgpool.patch]: Is this still required? We have glibc 2.27 by default now

**Comment by danbst** [comment]
As for patch itself - wouldn't it be more safe to use absolute path only if `pool_config->pool_passwd` starts with `/`, and resolve relative in other cases? 

In general, we avoid patches which change logic too much, and this change requires package documentation changes. This restriction can be lifted if patch is submitted upstream.

File: pkgs/servers/sql/pgpool/pgpool.patch (hunk 1 above)
Comment (takeda) [reply pkgs/servers/sql/pgpool/pgpool.patch]: When I was upgrading, I started without any patches and it failed to compile. Perhaps my version of nixpkgs was slightly outdated? When did that change?

**Comment by takeda** [comment]
@danbst that's a good suggestion, will try that

**Comment by takeda** [comment]
@danbst Looks like the patch around watchdog is indeed no longer needed so I removed it (I must have old version of nixpkgs) I also applied your suggestion to treat option as an absolute path if it starts with a slash.

**Comment by danbst** [comment]
@takeda I've proposed patch upstream http://www.sraoss.jp/pipermail/pgpool-hackers/2019-August/003366.html. Let's wait a bit for feedback, but overall I think this is fine and can be merged.

**Comment by danbst** [comment]
@takeda thanks for your contribution! The patch would most likely be included upstream, so I went ahead and merged it.

**Comment by takeda** [comment]
@danbst thank you

**Pull request merged**